src/compiler/path.ts

```diff
@@ -461,12 +461,15 @@ namespace ts {
      */
     export function reducePathComponents(components: readonly string[]) {
         if (!some(components)) return [];
-        const reduced = [components[0]];
+        let reduced: string[] | undefined;
         for (let i = 1; i < components.length; i++) {
             const component = components[i];
-            if (!component) continue;
-            if (component === ".") continue;
+            if (!component || component === ".") {
+                reduced = reduced ?? components.slice(0, i);

```

Comment:
I'd just use `||` here, but your preference.

Reply:
What's the difference?

Reply:
For arrays and other object types, they're the same, but the `??` emit is more verbose.

Reply:
Cool, if I don't abandon this, I'll change it.